File: sdk/go/provider/fakemail.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	stdhttp "net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

const fakemailBase = "https://www.fakemail.net"

var fakemailCSRFRe = regexp.MustCompile(`const\s+CSRF\s*=\s*"([^"]+)"`)

type fakemailGenerateResponse struct {
	Email string `json:"email"`
	Heslo string `json:"heslo"`
}

type fakemailListRow struct {
	ID       any    `json:"id"`
	Subject  string `json:"predmet"`
	From     string `json:"od"`
	When     string `json:"kdy"`
	ReadFlag string `json:"precteno"`
}

type fakemailDetailResponse struct {
	ID      any    `json:"id"`
	Subject string `json:"predmet"`
	From    string `json:"od"`
	HTML    string `json:"telo"`
	To      string `json:"komu"`
}

func fakemailClient() *stdhttp.Client {
	return &stdhttp.Client{Timeout: 15 * time.Second}
}

func fakemailMergeCookie(prev string, headers stdhttp.Header) string {
	jar := map[string]string{}
	for _, part := range strings.Split(prev, ";") {
		p := strings.TrimSpace(part)
		if i := strings.Index(p, "="); i > 0 {
			jar[p[:i]] = p[i+1:]
		}
	}
	for _, line := range headers.Values("Set-Cookie") {
		first := strings.TrimSpace(strings.Split(line, ";")[0])
		if i := strings.Index(first, "="); i > 0 {
			jar[first[:i]] = first[i+1:]
		}
	}
	parts := make([]string, 0, len(jar))
	for k, v := range jar {
		parts = append(parts, k+"="+v)
	}
	return strings.Join(parts, "; ")
}

func fakemailRequest(method, u, cookie string, body io.Reader) (int, []byte, string, error) {
	req, err := stdhttp.NewRequest(method, u, body)
	if err != nil {
		return 0, nil, "", err
	}
	req.Header.Set("User-Agent", "Mozilla/5.0")
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
	if strings.Contains(u, "/index/") {
		req.Header.Set("Accept", "application/json, text/javascript, */*; q=0.01")
		req.Header.Set("X-Requested-With", "XMLHttpRequest")
		req.Header.Set("Referer", fakemailBase+"/")
	}
	if method == stdhttp.MethodPost {
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	}
	if cookie != "" {
		req.Header.Set("Cookie", cookie)
	}
	resp, err := fakemailClient().Do(req)
	if err != nil {
		return 0, nil, "", err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return resp.StatusCode, nil, "", err
	}
	return resp.StatusCode, data, fakemailMergeCookie(cookie, resp.Header), nil
}

func fakemailCleanJSON(data []byte) []byte {
	if len(data) >= 3 && data[0] == 0xef && data[1] == 0xbb && data[2] == 0xbf {
		return data[3:]
	}
	return data
}

func FakemailGenerate() (*CreatedMailbox, error) {
	status, body, cookie, err := fakemailRequest(stdhttp.MethodGet, fakemailBase+"/", "", nil)
	if err != nil {
		return nil, err
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("fakemail home: %d", status)
	}
	m := fakemailCSRFRe.FindStringSubmatch(string(body))
	if len(m) < 2 {
		return nil, fmt.Errorf("fakemail: csrf token not found")
	}
	u := fakemailBase + "/index/index?csrf_token=" + url.QueryEscape(m[1])
	status, body, cookie, err = fakemailRequest(stdhttp.MethodGet, u, cookie, nil)
	if err != nil {
		return nil, err
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("fakemail generate: %d", status)
	}
	var data fakemailGenerateResponse
	if err := json.Unmarshal(fakemailCleanJSON(body), &data); err != nil {
		return nil, err
	}
	email := strings.TrimSpace(data.Email)
	if email == "" || !strings.Contains(email, "@") {
		return nil, fmt.Errorf("fakemail: invalid mailbox response")
	}
	return &CreatedMailbox{Channel: "fakemail", Email: email, Token: cookie}, nil
}

func fakemailFetchDetail(cookie, id string) (*fakemailDetailResponse, error) {
	form := url.Values{"id": {id}}
	status, body, _, err := fakemailRequest(stdhttp.MethodPost, fakemailBase+"/index/email", cookie, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("fakemail detail: %d", status)
	}
	var detail fakemailDetailResponse
	if err := json.Unmarshal(fakemailCleanJSON(body), &detail); err != nil {
		return nil, err
	}
	return &detail, nil
}

func fakemailFlatten(row fakemailListRow, detail *fakemailDetailResponse, recipient string) map[string]any {
	subject := row.Subject
	from := row.From
	html := ""
	id := row.ID
	if detail != nil {
		if detail.Subject != "" {
			subject = detail.Subject
		}
		if detail.From != "" {
			from = detail.From
		}
		if detail.HTML != "" {
			html = detail.HTML
		}
		if detail.ID != nil {
			id = detail.ID
		}
	}
	return map[string]any{
		"id":      id,
		"from":    from,
		"to":      recipient,
		"subject": subject,
		"text":    "",
		"html":    html,
		"date":    row.When,
		"isRead":  row.ReadFlag == "precteno",
	}
}

func FakemailGetEmails(token, email string) ([]NormEmail, error) {
	cookie := strings.TrimSpace(token)
	address := strings.TrimSpace(email)
	if cookie == "" {
		return nil, fmt.Errorf("fakemail: empty session token")
	}
	if address == "" {
		return nil, fmt.Errorf("fakemail: empty email")
	}
	status, body, _, err := fakemailRequest(stdhttp.MethodGet, fakemailBase+"/index/refresh", cookie, nil)
	if err != nil {
		return nil, err
	}
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("fakemail refresh: %d", status)
	}
	var rows []fakemailListRow
	if err := json.Unmarshal(fakemailCleanJSON(body), &rows); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		id := strings.TrimSpace(fmt.Sprintf("%v", row.ID))
		var detail *fakemailDetailResponse
		if id != "" {
			detail, _ = fakemailFetchDetail(cookie, id)
		}
		out = append(out, NormalizeMap(fakemailFlatten(row, detail, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/fakemail.ts
import { setDefaultResultOrder } from 'dns';
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'fakemail';
const BASE_URL = 'https://www.fakemail.net';
const CSRF_RE = /const\s+CSRF\s*=\s*"([^"]+)"/;

interface GenerateResponse {
  email?: string;
  heslo?: string;
}

interface ListRow {
  id?: string | number;
  predmet?: string;
  od?: string;
  kdy?: string;
  precteno?: string;
}

interface DetailResponse {
  id?: string | number;
  predmet?: string;
  od?: string;
  telo?: string;
  komu?: string;
}

function preferIPv4(): void {
  try {
    setDefaultResultOrder('ipv4first');
  } catch {
    /* Older Node versions may not support this option. */
  }
}

function setCookieLines(headers: Headers): string[] {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === 'function') return h.getSetCookie();
  const one = headers.get('set-cookie');
  return one ? one.split(/,(?=[^ ;]+=)/) : [];
}

function mergeCookieHeader(prev: string, headers: Headers): string {
  const jar = new Map<string, string>();
  for (const part of prev.split(';')) {
    const p = part.trim();
    const i = p.indexOf('=');
    if (i > 0) jar.set(p.slice(0, i), p.slice(i + 1));
  }
  for (const line of setCookieLines(headers)) {
    const first = line.split(';')[0]?.trim() || '';
    const i = first.indexOf('=');
    if (i > 0) jar.set(first.slice(0, i), first.slice(i + 1));
  }
  return [...jar.entries()].map(([k, v]) => `${k}=${v}`).join('; ');
}

function parseJSON<T>(text: string): T {
  return JSON.parse(text.replace(/^\uFEFF/, '')) as T;
}

function homeHeaders(): Record<string, string> {
  return {
    Accept: 'text/html,application/xhtml+xml',
    'User-Agent': 'Mozilla/5.0',
  };
}

function ajaxHeaders(cookie: string): Record<string, string> {
  return {
    Accept: 'application/json, text/javascript, */*; q=0.01',
    'X-Requested-With': 'XMLHttpRequest',
    Referer: `${BASE_URL}/`,
    Cookie: cookie,
    'User-Agent': 'Mozilla/5.0',
  };
}

async function openSession(): Promise<{ csrf: string; cookie: string }> {
  preferIPv4();
  const response = await fetchWithTimeout(`${BASE_URL}/`, { headers: homeHeaders() });
  if (!response.ok) throw new Error(`fakemail home: ${response.status}`);
  const cookie = mergeCookieHeader('', response.headers);
  const html = await response.text();
  const csrf = CSRF_RE.exec(html)?.[1] || '';
  if (!csrf) throw new Error('fakemail: csrf token not found');
  return { csrf, cookie };
}

async function createAddress(session: { csrf: string; cookie: string }): Promise<{ data: GenerateResponse; cookie: string }> {
  preferIPv4();
  const response = await fetchWithTimeout(`${BASE_URL}/index/index?csrf_token=${encodeURIComponent(session.csrf)}`, {
    headers: ajaxHeaders(session.cookie),
  });
  if (!response.ok) throw new Error(`fakemail generate: ${response.status}`);
  const cookie = mergeCookieHeader(session.cookie, response.headers);
  const data = parseJSON<GenerateResponse>(await response.text());
  return { data, cookie };
}

async function listRows(cookie: string): Promise<ListRow[]> {
  preferIPv4();
  const response = await fetchWithTimeout(`${BASE_URL}/index/refresh`, {
    headers: ajaxHeaders(cookie),
  });
  if (!response.ok) throw new Error(`fakemail refresh: ${response.status}`);
  const rows = parseJSON<unknown>(await response.text());
  return Array.isArray(rows) ? rows as ListRow[] : [];
}

async function fetchDetail(cookie: string, id: string): Promise<DetailResponse | null> {
  preferIPv4();
  try {
    const response = await fetchWithTimeout(`${BASE_URL}/index/email`, {
      method: 'POST',
      headers: {
        ...ajaxHeaders(cookie),
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: new URLSearchParams({ id }).toString(),
    });
    if (!response.ok) return null;
    return parseJSON<DetailResponse>(await response.text());
  } catch {
    return null;
  }
}

function flatten(row: ListRow, detail: DetailResponse | null, recipient: string): Record<string, unknown> {
  return {
    id: detail?.id ?? row.id,
    from: detail?.od || row.od || '',
    to: recipient,
    subject: detail?.predmet || row.predmet || '',
    text: '',
    html: detail?.telo || '',
    date: row.kdy || '',
    isRead: row.precteno === 'precteno',
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const session = await openSession();
  const { data, cookie } = await createAddress(session);
  const email = String(data.email || '').trim();
  if (!email || !email.includes('@')) {
    throw new Error('fakemail: invalid mailbox response');
  }
  return { channel: CHANNEL, email, token: cookie };
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const cookie = String(token || '').trim();
  const address = String(email || '').trim();
  if (!cookie) throw new Error('fakemail: empty session token');
  if (!address) throw new Error('fakemail: empty email');
  const rows = await listRows(cookie);
  const emails: Email[] = [];
  for (const row of rows) {
    const id = String(row.id || '').trim();
    const detail = id ? await fetchDetail(cookie, id) : null;
    emails.push(normalizeEmail(flatten(row, detail, address), address));
  }
  return emails;
}
